### ISSUE TITLE
bind: cannot assign requested address

### ISSUE THREAD

**Issue by sebelk** [ibuildthecloud/systemd-docker#23]
Hi,

I have the following mydockerserv.service:

~~~
[Unit]
Description=mydockerserv
After=docker.service
Requires=docker.service

[Service]
ExecStart=/opt/bin/systemd-docker --cgroups name=systemd --cgroups=cpu run  -p X.W.Y.Z:443:443    --name %n sebelk/mydockerserv
Restart=always
RestartSec=10s
Type=notify
NotifyAccess=all
TimeoutStartSec=120
TimeoutStopSec=15

[Install]
WantedBy=multi-user.target
~~~

If I change the old IP address of the host A.B.C.D by  X.W.Y.Z, and then restart the service, it does not start the service, the error is:

~~~
Mar  9 07:23:43 newhope systemd: Unit mydockersev.service entered failed state.
Mar  9 07:23:43 newhope systemd: mydockerserv.service failed.
Mar  9 07:23:43 newhope systemd-docker: 2015/03/09 07:23:43 API error (500): Cannot start container d3e704d3135cfc99992f33d8da263bf79312fbb6f30aa8171dbb434363c4996d: Error starting userland proxy: listen tcp A.B.C.D:443: bind: cannot assign requested address
~~~

I've found that **only** removing the old container fixes the problem, any ideas?

**Comment by ibuildthecloud** [comment]
The port definition is stored in Docker.  It is best to always add "--rm" to your container.  If you don't then Docker will effectively cache the IP.  By adding "--rm" systemd-docker will ensure that the container is deleted and recreated on each start effectively clearing the saved IP

**Comment by sebelk** [comment]
Good point. But if I add "-rm" whenever stop/restart mydockerserv new data are gone. How can I prevent it?

**Comment by pward123** [comment]
You should use a docker data container.  https://medium.com/@ramangupta/why-docker-data-containers-are-good-589b3c6c749e

On a side note, I prefer to perform the rm function as part of the service startup instead of using --rm.  The apps in my containers send logging to stdout.  Keeping the old container around after the service is stopped makes it easier to perform postmortems.

Here's an example systemd:
```
[Unit]
Description=my-service
After=docker.service
Requires=docker.service
 
[Service]
TimeoutStartSec=0
ExecStartPre=-/usr/bin/docker kill my-service
ExecStartPre=-/usr/bin/docker rm my-service
 
ExecStart=/opt/bin/systemd-docker run \
    --name my-service \
    -p 8000:80 \
    -v /tmp/backups:/backups \
    --volumes-from my-data-volume \
    diem/my-service:1.0.2
 
ExecStop=/usr/bin/docker stop my-service
 
Restart=always
RestartSec=10s
Type=notify
NotifyAccess=all
TimeoutStartSec=240
TimeoutStopSec=15
 
[Install]
WantedBy=multi-user.target
```

I'm using two kinds of volume mounts.  The "backups" one allows the container to write to the docker host machine.  This allows you to do things later like docker exec mongodump to backup your mongo container using mongo tools.

The "my-data-volume" mount is using a docker data container.

**Comment by ibuildthecloud** [comment]
You need to store your data in volumes from a data only container.  So create another containers with the volumes you need and then do "--volumes-from"

**Comment by sebelk** [comment]
OK, I've got... but I'm trying to understand the rationale, for example let's say you have a container with apache, eg

1) I pull a distro image
2) I create a "Data only container" with /etc/httpd, /var/www and /var/log/httpd as volumes
3) Then I run with --volumes-from and install packages needed
4) I install apache et al
5) Should I commit the chanages to the container?

Please if you have a better idea there will be welcome :)
